packages/patrol_log/lib/src/patrol_log_reader.dart

```diff
@@ -0,0 +1,207 @@
+
+class PatrolLogReader {
+  PatrolLogReader({
+    required DisposeScope scope,
+    required this.listenStdOut,
+    required this.log,
+    required this.reportPath,
+    required this.showFlutterLogs,
+    required this.showTestSteps,
+  }) : _scope = scope;
+
+  final void Function(String) log;
+  final String reportPath;
+  final bool showFlutterLogs;
+  final bool showTestSteps;
+  final StreamSubscription<void> Function(
+    void Function(String) onData, {
+    Function? onError,
+    void Function()? onDone,
+    bool? cancelOnError,
+  }) listenStdOut;
+  final DisposeScope _scope;
+
+  final _stopwatch = Stopwatch();
+  final List<PatrolSingleTestEntry> _singleEntries = [];
+  final List<String> _skippedTests = [];
+  final StreamController<Entry> _controller =
+      StreamController<Entry>.broadcast();
+  late final StreamSubscription<Entry> _streamSubscription;
+
+  void listen() {
+    read();
+    listenStdOut(parse).disposedBy(_scope);
+  }
+
+  /// Starts the timer measuring whole tests duration.
+  void startTimer() => _stopwatch.start();
+
+  /// Stops the timer measuring whole tests duration.
+  void stopTimer() => _stopwatch.stop();
+
+  /// Returns the total number of tests.
+  int get totalTests => _singleEntries.length;
+
+  /// Returns the number of tests that passed.
+  int get successfulTests =>
+      _singleEntries.where((e) => e.status == TestEntryStatus.success).length;
+
+  /// Return list of failed tests.
+  List<PatrolSingleTestEntry> get failedTests =>
+      _singleEntries.where((e) => e.status == TestEntryStatus.failure).toList();
+
+  /// Returns the number of failed tests.
+  int get failedTestsCount => failedTests.length;
+
+  /// Returns the number of skipped tests.
+  int get skippedTests =>
+      _singleEntries.where((e) => e.status == TestEntryStatus.skip).length;
+
+  /// Returns `String` that have printed in bullet points name of failed tests
+  /// with relative path to file that contains the test.
+  String get failedTestsList =>
+      failedTests.map((e) => '  - ${e.nameWithPath}').join('\n');
+
+  void parse(String line) {
+    if (line.contains('PATROL_LOG')) {
+      final regExp = RegExp(r'PATROL_LOG \{(.*?)\}');
+      final match = regExp.firstMatch(line);
+      if (match != null) {
+        final matchedText = match.group(1)!;
+        final json = '{$matchedText}';
+        final entry = parseEntry(json);
+        if (entry case TestEntry _) {
+          final testEntry = entry;
+          // Skip info test is returned multiple times, so we need to filter it
+          if (testEntry.status == TestEntryStatus.skip &&
+              !_skippedTests.contains(testEntry.name)) {
+            _skippedTests.add(testEntry.name);
+            _controller.add(entry);
+          } else if (testEntry.status != TestEntryStatus.skip) {
+            _controller.add(entry);
+          }
+        } else {
+          _controller.add(entry);
+        }
+      }
```

Comment:
I would move this block of code to separate private method with a little explanation what is happening (and especially a name explaining what's happening inside)